src/decision/location_selection.jl

```diff
@@ -0,0 +1,265 @@
+    in_conn = in_conn[site_ids]
+    out_conn = out_conn[site_ids]
+    strong_pred = strong_pred[site_ids]
+
+    (_, _, ranks) = guided_site_selection(mcda_vars, scenario.guided, true, true, prefseedsites, prefshadesites, rankingsin, in_conn, out_conn, strong_pred)
+
+    return ranks[:, 2:3]
+end
+
+"""
+    run_site_selection(domain::Domain, scenarios::DataFrame, sum_cover::AbstractArray, area_to_seed::Float64; aggregate_function=nothing,
+        target_seed_sites=nothing, target_shade_sites=nothing)
+    run_site_selection(domain::Domain, scenarios::DataFrame, sum_cover::AbstractArray, area_to_seed::Float64, aggregation_function::Function;
+        target_seed_sites=nothing, target_shade_sites=nothing)
+
+Perform site selection for a given domain for multiple scenarios defined in a dataframe.
+
+# Arguments
+- `domain` : ADRIA Domain type, indicating geographical domain to perform site selection over.
+- `scenarios` : DataFrame of criteria weightings and thresholds for each scenario.
+- `sum_cover` : array of size (number of scenarios * number of sites) containing the summed coral cover for each site selection scenario.
+- `area_to_seed` : area of coral to be seeded at each time step in km^2
+- `aggregation_function` : function which aggregates ranks output into preferred form, e.g `ranks_to_frequencies`, `ranks_to_location_order`.
+- `target_seed_sites` : list of candidate locations for seeding (indices)
+- `target_shade_sites` : list of candidate location to shade (indices)
+
+# Returns
+- `ranks_store` : number of scenarios * sites * 3 (last dimension indicates: site_id, seed rank, shade rank)
+    containing ranks for each scenario run.
+"""
+function run_site_selection(domain::Domain, scenarios::DataFrame, sum_cover::AbstractArray, area_to_seed::Float64;
+    target_seed_sites=nothing, target_shade_sites=nothing)
+    ranks_store = NamedDimsArray(
+        zeros(length(domain.site_ids), 2, nrow(scenarios)),
+        sites=domain.site_ids,
+        intervention=1:2,
+        scenarios=1:nrow(scenarios),
+    )
+
+    dhw_scens = domain.dhw_scens
+    wave_scens = domain.wave_scens
+
+    # Pre-calculate maximum depth to consider
+    target_dhw_scens = unique(scenarios[:, "dhw_scenario"])
+    target_wave_scens = unique(scenarios[:, "wave_scenario"])
+
+    target_site_ids = Int64[]
+    if !isnothing(target_seed_sites)
+        append!(target_site_ids, target_seed_sites)
+    end
+
+    if !isnothing(target_shade_sites)
+        append!(target_site_ids, target_shade_sites)
+    end
+
+    if isnothing(target_seed_sites) && isnothing(target_shade_sites)
+        target_site_ids = collect(1:length(domain.site_ids))
+    end
+
+    n_sites = length(domain.site_ids)
+    for (scen_idx, scen) in enumerate(eachrow(scenarios))
+
+        depth_criteria = (domain.site_data.depth_med .<= (scen.depth_min .+ scen.depth_offset)) .& (domain.site_data.depth_med .>= scen.depth_min)
```

Comment:
You could assign `domain.site_data.depth_med .<= (scen.depth_min .+ scen.depth_offset)` and `domain.site_data.depth_med .>= scen.depth_min` to separate variables. The number of allocations would still be the same as well as the memory usage. I did the following test here:

```julia
function func1(x, y, z)
    iguais = x .== y
    return z[iguais]
end

function func2(x, y, z)
    return z[x .== y]
end

x, y, z = bitrand(100), bitrand(100), bitrand(100)

@btime func1(x,y,z)
#   135.755 ns (4 allocations: 208 bytes)
@btime func2(x,y,z)
#  136.114 ns (4 allocations: 208 bytes)
```

Reply:
This will all be done in an external function as in #434